src/rust/lib/graph-editor/src/component/edge.rs

```diff
@@ -44,23 +47,280 @@ const NODE_PADDING       : f32 = node::SHADOW_SIZE;
+enum Area {
+    Above,
+    Below,
+}
+
+/// Holds the data required to split a shape into two parts.
+///
+/// The `area` indicates which side of the split will receive special coloring.
+#[derive(Clone,Copy,Debug)]
+struct SplitData {
+    split_position : Vector2<f32>,
+    area           : Area,
+    cut_angle      : f32
+}
+
+impl SplitData {
+    fn new(split_position:Vector2<f32>, area: Area, cut_angle:f32) -> Self {
+        SplitData {split_position,area,cut_angle}
+    }
+}
+
+/// `SnapData` is used as the result value of snapping operations on `MultiShapes`. It holds the
+/// position and shape that a hover position was snapped to, that is the shape and the position
+/// on the shape that was closest to the hover
+struct SnapData<'a> {
+    position     : Vector2<f32>,
+    target_shape : &'a dyn EdgeShape,
+}
+
+impl<'a> SnapData<'a> {
+    fn new(position: Vector2<f32>, target_shape: &'a dyn EdgeShape) -> Self {
+        SnapData{position,target_shape}
+    }
+}
+
+/// The MultiShape trait allows operations on a collection of `EdgeShape`.
+trait MultiShape {
+    /// Return the `ShapeViewEvents` of all sub-shapes.
+    fn events(&self) -> Vec<ShapeViewEvents>;
+    /// Return references to all `EdgeShape`s in this MultiShape.
+    fn edge_shape_views(&self) -> Vec<&dyn EdgeShape>;
+
+    /// Connect the given `ShapeViewEventsProxy` to the mouse events of all sub-shapes.
+    fn register_proxy_frp(&self, network:&frp::Network, frp:&ShapeViewEventsProxy) {
+        let events = self.events();
+        for event in &events {
+            frp::extend! { network
+                eval_ event.mouse_down (frp.on_mouse_down.emit(()));
+                eval_ event.mouse_over (frp.on_mouse_over.emit(()));
+                eval_ event.mouse_out (frp.on_mouse_out.emit(()));
+            }
+        }
+    }
+
+    /// Snap the given position to our sub-shapes. Returns `None` if the given position cannot be
+    /// snapped to any sub-shape (e.g., because it is too far away).
+    fn snap_position_to_shape(&self, point:Vector2<f32>) -> Option<SnapData> {
+        let shapes    = self.edge_shape_views();
+        let snap_data = shapes.iter().filter_map(|shape|{
+            let maybe_snap_position     = shape.snap_to_shape(point);
+            if let Some(snap_position) = maybe_snap_position {
+                // TODO The `SNAP_THRESHOLD` test is needed to avoid visual artifacts for
+                // circles where sometimes we might snap to a part that is invisible instead of
+                // the visible shape next to it. It might be better to find a cheaper way to avoid
+                // these.
+                let snap_distance = (snap_position - point).norm();
+                if snap_distance < SNAP_THRESHOLD && shape.contains(point){
+                    return Some(SnapData::new(snap_position,*shape))
+                }
+            }
+            None
+        });
+
+        // Return the snap that is closest to the hover.
```

Comment:
We should not try to snap to all shapes. Instead, we should discover the correct shape from the hover event directly and only snap to that one. This will avoid also needing the `SNAP_THRESHOLD` in the first place.

Reply:
I would like to do both of these things as part of the next PR that has up and down edges working if that is fine with you.  It avoids some backporting/merging work